Fix For: The ':' character, hexadecimal value 0x3A, cannot be included in a name.

keith-hsv opened the pull request:
I found that FormatThicknessService's call to IsMatch in the ProcessElement method was just passing a string value of the attribute that was implicitly being converter to XName.

If an attribute value was an attached property, "local:MyProperty" then this would generate an exception.

I added an AttributeHelper static class that has a GetName method. GetName takes an XAttribute and returns a properly configured XName taking into account the namespace of an attached property.

My naming conventions might not be in-line with what you want, and I am definitely up for any changes to make it better.

Commit:
Updated ProcessElement's call to IsMatch to use the new AttributeHelper.GetName method that returns a properly configured XName object, taking into account namespaces of attached properties.

Commit:
Added new AttributeHelper static class with a new method GetName that will return a properly configured XName object from an XAttribute. GetName takes into account namespaces of attached properties.

grochocki commented:
Can you add a unit test to validate the change? Also, did you verify that you did not break any existing unit tests?

ghost commented:
I will do that right away and make sure that all unit tests pass. Stand by.

ghost commented:
All unit tests pass. I am working on the new unit test. I will add it over the weekend or on Monday.

grochocki commented:
Great, thanks!

ghost commented:
I am struggling with the .testxaml, .expected files in my unit test. 

The test keeps failing when it compares the files. My .expected file was run through XamlMagic but the formatting is always different from what the test expects.

Is there a trick to this?

ghost commented:
I figured out the .testxaml / .expected issue. :)

Commit:
Added test for attached property syntax in an attribute.

ghost commented:
I have added a test that will verify that attached property syntax is working.

grochocki commented:
Thanks! I will take this for spin and try to close it up this week.

ghost commented:
I'm definitely interested in any ways it could be better or different. Also, if you see there is a better way of testing what I did, please let me know so that I can learn. :)

BTW, it was really easy to get the solution up and running and debugging! Kudos!

punker76 commented:
@grochocki 

> I will take this for spin and try to close it up this week.

:+1:

grochocki commented on UnitTests/TestFiles/TestAttachedProperty.expected:
```diff
@@ -0,0 +1,9 @@
+﻿<ResourceDictionary
```
Insert spaces instead of tabs.

grochocki commented on UnitTests/UnitTests.AttachedProperty.cs:
```diff
@@ -0,0 +1,20 @@
+        {
+            var stylerOptions = new StylerOptions(config: this.defaultConfig)
+			{
+				IndentWithTabs = true
```
You should be able to remove this if you use spaces in the .expected file.

grochocki commented on UnitTests/UnitTests.csproj:
```diff
@@ -36,6 +36,9 @@
     <Prefer32Bit>false</Prefer32Bit>
   </PropertyGroup>
   <ItemGroup>
+    <Reference Include="Microsoft.VisualStudio.QualityTools.UnitTestFramework, Version=10.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL">
```
Is this ref update necessary?

keith-hsv replied:
Not sure. Visual Studio must of done this automatically.

grochocki commented on Service/Reorder/FormatThicknessService.cs:
```diff
@@ -39,14 +39,14 @@ public void ProcessElement(XElement element)
+					if (valueAttribute != null)
+					{
+						this.FormatAttribute(valueAttribute);
+					}
```
Replace tabs with spaces, which will reduce the number of changes here. You can change this setting in Tools --> Options --> Text Editor --> All Languages --> Tabs. In general, use 4 spaces instead of a tab.

Edit: Do this for all touched lines. You can also visualize spaces vs. tabs by going to Edit --> Advanced --> View White Space.

grochocki commented on Service/Helpers/AttributeHelper.cs:
```diff
@@ -0,0 +1,32 @@
+					string prefix = parts[0];
+					string value = parts[1];
+
+					var ns = attribute.Parent.GetNamespaceOfPrefix(prefix);
```
I know there are violations of this all over, so don't worry about this one, but in general, avoid abbreviations, as they are more difficult to parse to someone not as familiar with the code.

grochocki commented:
Verified functionality. A few nitpicky comments, but otherwise looks good. Fix the tabbing, and it should be good to go.

Commit:
Updated to use spaces vs. tabs.
Updated AttributeHelper to use a full variable name vs. an abbreviation.
Updated unit test to use the legacyConfig.

ghost commented:
Updated code based on your direction. All unit tests passed.

grochocki commented:
Looks good, thanks!
